src/doip_simulator.cpp

```diff
@@ -0,0 +1,188 @@
+ */
+void DoIPSimulator::listenUdp() {
+    doipserver->setupUdpSocket();
+    while(1) {
+        doipserver->receiveUdpMessage();
+    }
+}
+
+/*
+ * Check permantly if tcp message was received
+ */
+void DoIPSimulator::listenTcp() {
+    doipserver->setupSocket();
+    while(1) {
+        doipserver->receiveMessage();
+    }
+}
+
+/**
+ * Is called when the doip library receives a diagnostic message.
+ * @param address   logical address of the ecu
+ * @param data      message which was received
+ * @param length    length of the message
+ */
+void DoIPSimulator::receiveFromLib(unsigned char* address, unsigned char* data, int length) {
+    printf("CarSimulator DoIP Simulator received: ");
+    for(int i = 0; i < length; i++) {
+        printf("0x%02X ", data[i]);
+    }
+    printf(" from doip lib.\n");
+    
+    int index = findECU(address);
+    if(index != -1) {
+        std::vector<unsigned char> response = ecus.at(index)->getUdsReceiver()->proceedDoIPData(data,length);
+        
+        if(response.size() > 0) {
+            unsigned char logicalAddress [2] = {ecus.at(index)->getLogicalAddress() >> 8, ecus.at(index)->getLogicalAddress() & 0xFF };
+            sendDiag(response, logicalAddress);
+        }
+    }
+}
+
+/**
+ * Passes the received response from a ecu back to the doip library
+ * @param data              respone from a ecu that will be send back
+ * @param logicalAddress    logical address of the ecu where data came from
+ */
+void DoIPSimulator::sendDiag(const std::vector<unsigned char> data, unsigned char* logicalAddress) {
+    
+    unsigned char* msg = new unsigned char[data.size()];
+    for(unsigned int i = 0; i < data.size(); i++) {
+        msg[i] = data[i];
+    }
+    
+    doipserver->receiveDiagnosticPayload(logicalAddress, msg, data.size());
+    delete[] msg;
+}
+
+/**
+ * Adds a ECU to the list
+ * @param ecu   Pointer to the ecu
+ */
+void DoIPSimulator::addECU(ElectronicControlUnit* ecu) {
+    ecus.push_back(ecu);
+}
+
+/**
+ * Will be called when the doip library receives a diagnostic message.
+ * The library notifies the application about the message.
+ * Checks if there is a ecu with the logical address
+ * @param targetAddress     logical address to the ecu
+ * @return                  If a positive or negative ACK should be send to the client
+ */
+bool DoIPSimulator::diagMessageReceived(unsigned char* targetAddress) {
+    PayloadType ackType;
+    unsigned char ackCode;
+    
+    //if there isnt a ecu with the target address 
+    if(findECU(targetAddress) == -1) {
+        //send negative ack with unknown target address and return
+        ackType = PayloadType::DIAGNOSTICNEGATIVEACK;
+        ackCode = 0x03;
+        doipserver->sendDiagnosticAck(ackType, ackCode);
+        return false;
+    }
+  
+    //positiv ack
+    ackType = PayloadType::DIAGNOSTICPOSITIVEACK;
+    ackCode = 0x00;
+    
+    //negative ack
+    //ackType = PayloadType::DIAGNOSTICNEGATIVEACK;
+    //ackCode = 0x02;
```

Comment:
Why is this part with the negative acknowledgement commented out? Is this intended/wanted?

Reply:
This was inserted to test the method but we never fully removed it. Thank you for the hint, we removed some commented code.